File: Pb1/src/main/java/repository/dataBaze/RepositoryDBClient.java
package repository.dataBaze;
import model.Client;
import model.Entity;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JDBCUtils;

import repository.RepositoryClient;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;


public class RepositoryDBClient implements RepositoryClient {

    private final JDBCUtils databaseUtils;
    private static final Logger logger = LogManager.getLogger();

    public RepositoryDBClient(Properties properties) {

        logger.info("Initialising RaceDBRepository with properties: {}", properties);
        databaseUtils = new JDBCUtils(properties);
    }


    @Override
    public Client findOne(Integer integer) {
        logger.traceEntry();
        Connection connection = databaseUtils.getConnection();
        Client client = null;
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from " + " where id=?")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    var id = resultSet.getInt("id");
                    var clientUsername = resultSet.getString("Username");
                    var clientPassword = resultSet.getString("Password");
                    var clientName = resultSet.getString("ClientName");
                    var clientAddress = resultSet.getString("Address");

                    client = new Client(clientUsername,clientPassword,clientName,clientAddress);
                    client.setId(id);
                }
                logger.traceExit("Found 1 instance");
            } catch (SQLException e) {
                e.printStackTrace();
            }
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return client;
    }

    @Override
    public Iterable<Client> findAll() {
        logger.traceEntry("Finding task");
        Connection connection = databaseUtils.getConnection();
        List<Client> clients = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Client")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    var id = resultSet.getInt("id");
                    var clientUsername = resultSet.getString("Username");
                    var clientPassword = resultSet.getString("Password");
                    var clientName = resultSet.getString("ClientName");
                    var clientAddress = resultSet.getString("Address");

                    Client client = new Client(clientUsername,clientPassword,clientName,clientAddress);
                    client.setId(id);
                    clients.add(client);
                }
                logger.traceExit("Found {} instances", clients.size());
            }
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return clients;
    }


    @Override
    public Client save(Client entity) {
        logger.traceEntry("Saving client {}", entity);
        Connection connection = databaseUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into Client (Userneme, Password,NameClient, Address) values (?, ?, ?, ?)")) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.setString(3, entity.getNameClient());
            preparedStatement.setString(4, entity.getAddressClient());

            int result = preparedStatement.executeUpdate();
            logger.traceExit("Saved {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return entity;
    }

    @Override
    public Client delete(Integer integer) {
        logger.traceEntry("Deleting task {}", integer);
        Connection connection = databaseUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from Client where id=?")) {
            preparedStatement.setInt(1, integer);
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Client update(Client entity) {
        logger.traceEntry("Updating task {}", entity);
        Connection connection = databaseUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("update  set UserName =?, Password =?,NameClient =?, Address =? where id=?")) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.setString(3, entity.getNameClient());
            preparedStatement.setString(4, entity.getAddressClient());

            int result = preparedStatement.executeUpdate();
            logger.traceExit("Updated {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return entity;
    }

    private Client buildRefereeFromResultSet(ResultSet resultSet) throws SQLException {
        Client client  = null;
        var id = resultSet.getInt("id");
        var clientUsername = resultSet.getString("Username");
        var clientPassword = resultSet.getString("Password");
        var clientName = resultSet.getString("ClientName");
        var clientAddress = resultSet.getString("Address");

        client = new Client(clientUsername,clientPassword,clientName,clientAddress);
        client.setId(id);
        return client;
    }

    public Client findByUsername(String username) {
        logger.traceEntry("Finding task {}", username);
        Connection connection = databaseUtils.getConnection();
        Client referee = null;
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Client where Username=?")) {
            preparedStatement.setString(1, username);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    referee = buildRefereeFromResultSet(resultSet);
                }
                logger.traceExit("Found 1 instance");
            }
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return referee;
    }
}

File: Pb1/p1/p1/Repository/DB/RepositoryDBClient.cs
using System.Collections.Generic;
using System.Data;
using p1.Model;
using p1.Utils;
using log4net;
namespace p1.Repository.DB;

public class RepositoryDBClient : RepositoryClient
{
    private static readonly ILog logger = LogManager.GetLogger("RepositoryBDClient");
    private IDictionary<string, string> properties;
    
    public RepositoryDBClient(IDictionary<string, string> properties, Client client) : base(client)
    {
        logger.Info("Creating RepositoryBDClient");
        this.properties = properties;
    }



    public override Client FindOne(int id)
    {
        logger.InfoFormat("Entering FindOne with value {0}", id);
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from Client where id=@id";
            var parameterId = command.CreateParameter();
            parameterId.ParameterName = "@id";
            parameterId.Value = id;
            command.Parameters.Add(parameterId);
            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    int _id = dataReader.GetInt32(0);
                    string username = dataReader.GetString(1);
                    string password = dataReader.GetString(2);
                    string clientName = dataReader.GetString(3);
                    string address = dataReader.GetString(4);
                    var client = new Client(username,password,clientName,address);
                    client.Id = _id;
                    logger.InfoFormat("Exiting FindOne with value {0}", client);
                    return client;
                }
            }
        }
        logger.InfoFormat("Exiting FindOne with value {0}", null);
        return null;
    }
    
    public override IEnumerable<Client> FindAll()
    {
        logger.Info("Entering FindAll");
        IDbConnection connection = DbUtils.GetConnection(properties);
        IList<Client> clients = new List<Client>();
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select * from Client";
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    int _id = dataReader.GetInt32(0);
                    string username = dataReader.GetString(1);
                    string password = dataReader.GetString(2);
                    string clientName = dataReader.GetString(3);
                    string address = dataReader.GetString(4);
                    var client = new Client(username,password,clientName,address);
                    client.Id = _id;
                    clients.Add(client);
                }
            }
        }
        logger.Info("Exiting FindAll");
        return clients;
    }
    
    public override Client Save(Client entity)
    {
        logger.InfoFormat("Entering Save with value {0}", entity);
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "insert into Client (Username, Password, ClientName, Address) values (@Username, @Password,  @ClientName, @Address)";
            
            var username  = command.CreateParameter();
            username.ParameterName = "@Username";
            username.Value = entity.Username;
            command.Parameters.Add(username);
            
            var password = command.CreateParameter();
            password.ParameterName = "@Password";
            password.Value = entity.Password;
            command.Parameters.Add(password);
            
            var nameClient  = command.CreateParameter();
            nameClient.ParameterName = "@ClientName";
            nameClient.Value = entity.NameClient;
            command.Parameters.Add(nameClient);
            
          
            var addressParameter = command.CreateParameter();
            addressParameter.ParameterName = "@Address";
            addressParameter.Value =  entity.Address;
            command.Parameters.Add(addressParameter);
            
         
            
            var result = command.ExecuteNonQuery();
            logger.InfoFormat("Exiting Save with value {0}", result);
        }
        return entity;
    }
    
    public override int Delete(int id)
    {
        logger.InfoFormat("Entering Delete with value {0}", id);
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "delete from Client where id=@id";
            var idParameter = command.CreateParameter();
            idParameter.ParameterName = "@id";
            idParameter.Value = id;
            command.Parameters.Add(idParameter);
            var result = command.ExecuteNonQuery();
            logger.InfoFormat("Exiting Delete with value {0}", result);
        }
        return id;
    }
    
    public override Client Update(Client entity)
    {
        logger.InfoFormat("Entering Update with value {0}", entity);
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "update Client set Username=@Username, Password=@Password, ClientName=@ClientName, Address=@Address where id=@id";
            var username  = command.CreateParameter();
            username.ParameterName = "@Username";
            username.Value = entity.Username;
            command.Parameters.Add(username);
            
            var password = command.CreateParameter();
            password.ParameterName = "@Password";
            password.Value = entity.Password;
            command.Parameters.Add(password);
            
            var nameClient  = command.CreateParameter();
            nameClient.ParameterName = "@ClientName";
            nameClient.Value = entity.NameClient;
            command.Parameters.Add(nameClient);
            
          
            var addressParameter = command.CreateParameter();
            addressParameter.ParameterName = "@Address";
            addressParameter.Value =  entity.Address;
            command.Parameters.Add(addressParameter);

            var result = command.ExecuteNonQuery();
            logger.InfoFormat("Exiting Update with value {0}", result);
        }
        return entity;
    }
} 
